src/uu/tail/src/tail.rs

```diff
@@ -356,49 +636,543 @@ pub fn uu_app<'a>() -> Command<'a> {
+        // TODO: [2022-06; jhscheer] maybe use `--max-unchanged-stats` here to reduce fstat calls?
+        if settings.use_polling && settings.follow.is_some() {
+            for path in &files
+                .map
+                .keys()
+                .filter(|p| p.is_tailable())
+                .map(|p| p.to_path_buf())
+                .collect::<Vec<PathBuf>>()
+            {
+                if let Ok(new_md) = path.metadata() {
+                    let pd = files.map.get(path).unwrap();
+                    if let Some(old_md) = &pd.metadata {
+                        if old_md.is_tailable()
+                            && new_md.is_tailable()
+                            && old_md.got_truncated(&new_md)?
+                        {
+                            show_error!("{}: file truncated", pd.display_name.display());
+                            files.update_metadata(path, Some(new_md));
+                            files.update_reader(path)?;
+                        }
+                    }
+                }
+            }
+        }
 
-        let pid_is_dead = !read_some && settings.pid != 0 && process.is_dead();
-        read_some = false;
+        // with  -f, sleep for approximately N seconds (default 1.0) between iterations;
+        let rx_result = rx.recv_timeout(settings.sleep_sec);
+        if rx_result.is_ok() {
+            _event_counter += 1;
+            _timeout_counter = 0;
+        }
 
-        for (i, (reader, filename)) in readers.iter_mut().enumerate() {
-            // Print all new content since the last pass
+        match rx_result {
+            Ok(Ok(event)) => {
+                // eprintln!("=={:=>3}=====================dbg===", _event_counter);
+                // dbg!(&event);
+                // dbg!(files.map.keys());
+                // dbg!(&files.last);
+                // dbg!(&orphans);
+                // eprintln!("=={:=>3}=====================dbg===", _event_counter);
+                handle_event(&event, files, settings, &mut watcher, &mut orphans)?;
+            }
+            Ok(Err(notify::Error {
+                kind: notify::ErrorKind::Io(ref e),
+                paths,
+            })) if e.kind() == std::io::ErrorKind::NotFound => {
+                if let Some(event_path) = paths.first() {
+                    if files.map.contains_key(event_path) {
+                        let _ = watcher.unwatch(event_path);
+                    }
+                }
+            }
+            Ok(Err(notify::Error {
+                kind: notify::ErrorKind::MaxFilesWatch,
+                ..
+            })) => crash!(1, "{} resources exhausted", text::BACKEND),
+            Ok(Err(e)) => crash!(1, "{:?}", e),
+            Err(mpsc::RecvTimeoutError::Timeout) => {
+                _timeout_counter += 1;
+            }
+            Err(e) => crash!(1, "RecvError: {:?}", e),
+        }
+
+        // main print loop
+        for path in files.map.keys().cloned().collect::<Vec<_>>() {
```

Comment:
Why do you collect it into a `Vec` only to iterate over it?

Reply:
Yes, this is a bit ugly. I do it to work around `cannot borrow *files as mutable because it is also borrowed as immutable`. At the moment I have no better solution here.

Reply:
Oh I see, I thought `cloned` would be enough, but that leaves the borrow until all elements have been cloned of course, so that's indeed not enough